app/script/entity/Conversation.js

```diff
@@ -335,16 +335,24 @@ z.entity.Conversation = class Conversation {
 
   /**
    * Adds a single message to the conversation.
-   * @param {z.entity.Message} message_et - Message entity to be added to the conversation
-   * @returns {undefined} No return value
+   * @param {z.entity.Message} messageEntity - Message entity to be added to the conversation.
+   * @param {boolean} replaceDuplicate - If a duplicate already exists, replace it with the new entity.
+   * @returns {undefined} No return value.
    */
-  add_message(message_et) {
-    message_et = this._checkForDuplicate(message_et);
-    if (message_et) {
-      this.update_timestamps(message_et);
-      this.messages_unordered.push(message_et);
-      amplify.publish(z.event.WebApp.CONVERSATION.MESSAGE.ADDED, message_et);
+  add_message(messageEntity, replaceDuplicate = false) {
+    const duplicateEntity = this._findDuplicate(messageEntity);
+    this.update_timestamps(messageEntity);
+    if (duplicateEntity && !replaceDuplicate) {
+      // If there is a duplicate in db, but we don't want to replace it our job here is done.
+      return;
     }
+    if (duplicateEntity) {
+      const duplicateIndex = this.messages_unordered.indexOf(duplicateEntity);
+      this.messages_unordered.splice(duplicateIndex, 1, messageEntity);
+      return;
+    }
```

Comment:
We also still need the  existential check on the message entity here if I am not mistaken.  `update_timestamps()` and the others were previously safeguarded by this and now isn't anymore.

Please also move the early return into the `if(duplicateEntity)` check instead of doing the same thing twice.